README.rst
Introduction
============

The Relex project using Plone as a backend.

Features
========

* Exposes all projects to the public.
* Provide a web backoffice
* Store data in http://bruxelles.irisnet.be/relex




CHANGES.txt
Changelog
=========

1.0.15 (unreleased)
-------------------

- Nothing changed yet.


1.0.14 (2014-12-15)
-------------------

- Fix: organisation type in organisation vocab.


1.0.13 (2014-11-20)
-------------------

- Fix: check contact is not None in getContactOrganisation.
- Fix: translate legends in project edit.


1.0.12 (2014-11-20)
-------------------

- Fix: add missing dutch translations.
- Fix: give Site administrator permission for inactive/archived projects.


1.0.11 (2014-10-13)
-------------------

- Fix: Related objects at creation.
- Fix: Fix workflow for inactive and archived projects.

1.0.10 (2014-10-10)
-------------------

- Fix: always store id directly for related objects.
- Reword:
  - Cell -> Unité administrative/Direction
  - Organisation -> Administration/Organisme
  - Partenaire bruxellois -> Administration/Partenaire bruxellois.
- Fix: Show all info for brussels partners.
- Fix: Sort select input in search page
- Fix: Hide inactive projects in search page
- Fix: Allow search by name of all languages for keywords/themes for global search in search page

1.0.9 (2014-09-18)
------------------

- Fix some CSS due to collective.js.bootstrap


1.0.8 (2014-09-18)
------------------

- fix: project broken.


1.0.7 (2014-09-03)
------------------

- Fix(admin): IHM Continent : Titre de la table à corriger par Continent en lieu et place de région.
- Fix(admin): IHM Country : La liste des continents, remplir la liste avec la description en lieu et place du code
- Fix(admin) IHM Organisation : Si je sélectionne l'organisation suivante : [AEE] ADMINISTRATION DE L'ECONOMIE ET DE L'EMPLOI (MINISTERE DE LA REGION DE BRUXELLES-CAPITALE), il est impossible d'ajouter des cellules pour une organisation qui n'en possédait pas avant ??
- Fix(admin) IHM Organisation : Liste type d'organisation, il faut la remplir avec la description et pas le code du type d’organisation
- Fix(admin) IHM Contact :Liste Organisation, Cellule et fonction sont vides ? -> Non
- Fix (admin) IHM Contact: Il faut remplir les listes Organisation, cellule et fonction avec la description pas le code. -> ATTENTION BCP DE FONCTION N ONT PAS DE DESCRIPTION
- Fix (admin) IHM BrusselsPatners: Liste Organisation et cellule sont vides
- Fix (admin) IHM BrusselsPatners: Il faut remplir les listes Organisation et cellule avec la description pas le code.

- Feature(admin) : Pour tous les champs de type texte ont doit forcer la mise en majuscule
- Fix(admin): add feedback on save project button
- Fix(search): hide code input to anonymous
- Feature(search): add search by organisation (remove search by contact)
- feature(admin/contact): add filter of cells based on selected orgnisation
- feature(admin/cells): Associate an organisation with a cell from here
- feature(admin/settings): add feedback on all save buttons
- fix(admin/settings/*): cancel button on new vocab term
- feature(search): add search by partner's organisation
- fix(data): relations data should not have list in their values
- feature(admin/brusselspartners): add filter of cells based on selected orgnisation
- feature(admin/cell): add ability to see/remove/add organisations

1.0.6 (2014-08-18)
------------------

- Make setuphandler.


1.0.5 (2014-08-14)
------------------

- Improve css for projects detail and print page.


1.0.4 (2014-08-13)
------------------

- Add multi-selects on the search page for country, region and city.
- Fix contact layout on project detail print page.
- Add dutch translations.


1.0.3 (2014-08-07)
------------------

- Nothing changed yet.


1.0.2 (2014-07-14)
------------------

- Change translation files to utf-8.


1.0.1 (2014-07-11)
------------------

- Update styles to integrate product in the prb website.
- Hide status filter on the search page when user is not logged in.
- Update translations.
- Fix bug with search and special characters.
- Add security check to admin interface.


1.0 (2014-07-11)
----------------

- Initial release.


scripts/import_csv.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Script used to the CSVs in relex_web.
THIS WILL REMOVE EXISTING VOCABULARIES.
Launch as ./bin/instance run scripts/import_csv.py path_to_csv_folder
"""
import csv
import json
import os
import sys

from DateTime import DateTime
from zope.component.hooks import setSite
from zope.container.interfaces import INameChooser

from cirb.relex.browser.vocabulary import KEY_STORAGE

CSVS = (
    'CELLS',
    'CELLS_ORAGANIZATIONS',
#    'CITIES_BACKUP',
    'CITIES',
    'CONTACTS',
    'CONTINENTS',
    'COUNTRIES',
    'FUNCTIONS',
    'KEYWORDS',
    'KEYWORDS_THEMES',
#    'KEYWORDS_TMP',
    'ORGANIZATIONS',
    'ORGANIZATIONTYPES',
#    'PROJECTS_CITIES_BACKUP',
    'PROJECTS_CITIES',
    'PROJECTS_CONTACTS',
    'PROJECTS_COUNTRIES',
    'PROJECTS',
    'PROJECTS_REGIONS',
    'PROJECTSTATUSES',
    'REGIONS',
    'RELATIONTYPES',
    'THEMES',
#    'THEMES_TMP',
    )


def get_csv_dict(folder):
    csv_dict = {}
    for csv_name in CSVS:
        path = os.path.join(folder, "{}.csv".format(csv_name))
        with open(path, 'r') as csv_file:
            reader = csv.reader(csv_file)
            # Remove header
            reader.next()
            csv_dict[csv_name] = []
            for row in reader:
                csv_dict[csv_name].append(row)
    return csv_dict


# ================= CONTINENTS COUNTRIES REGIONS CITIES =============

def _import_continents(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": cell[8].decode('latin-1'),
            "description": {
                "fr": cell[4].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[3].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.continent'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported continents.')


def _import_countries(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": cell[5].decode('latin-1'),
            "continent": cell[9].decode('latin-1'),
            "name": {
                "fr": cell[2].decode('latin-1'),
                "en": cell[7].decode('latin-1'),
                "nl": cell[4].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.country'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported countries.')


def _import_regions(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "country": cell[8].decode('latin-1'),
            "name": {
                "fr": cell[4].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[5].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.region'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported regions.')


def _import_cities(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "country": cell[9].decode('latin-1'),
            "region": cell[8].decode('latin-1'),
            "name": {
                "fr": cell[3].decode('latin-1'),
                "en": cell[1].decode('latin-1'),
                "nl": cell[6].decode('latin-1'),
                },
            }
        terms.append(term)
    key = KEY_STORAGE + '.city'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported cities.')


# ========= CELLS ORGANIZATIONS RELATION FUNCTIONS CONTACTS  =================

def _import_cells(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": {
                "fr": cell[9].decode('latin-1'),
                "en": cell[10].decode('latin-1'),
                "nl": cell[5].decode('latin-1'),
                },
            "description": {
                "fr": cell[4].decode('latin-1'),
                "en": cell[7].decode('latin-1'),
                "nl": cell[3].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.cell'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported cells.')


def _import_relation_types(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": {
                "fr": cell[1].decode('latin-1'),
                "en": cell[3].decode('latin-1'),
                "nl": cell[2].decode('latin-1'),
                },
            "name": {
                "fr": cell[4].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[5].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.relationtype'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported relation types.')


def _import_organisation_types(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": {
                "fr": cell[1].decode('latin-1'),
                "en": cell[10].decode('latin-1'),
                "nl": cell[5].decode('latin-1'),
                },
            "name": {
                "fr": cell[4].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[2].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.organisationtype'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported organisation types.')


def _import_organisations(cells, cells_orgs, relex_web):
    cells_orgs_dict = {}
    for row in cells_orgs:
        if row[1] not in cells_orgs_dict.keys():
            cells_orgs_dict[row[1]] = [row[0]]
        else:
            cells_orgs_dict[row[1]].append([row[0]])
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "organisationtype": cell[12].decode('latin-1'),
            "url": cell[1].decode('latin-1'),
            "cell": cells_orgs_dict.get(cell[0], []),
            "code": {
                "fr": cell[5].decode('latin-1'),
                "en": cell[9].decode('latin-1'),
                "nl": cell[4].decode('latin-1'),
                },
            "name": {
                "fr": cell[11].decode('latin-1'),
                "en": cell[8].decode('latin-1'),
                "nl": cell[2].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.organisation'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported organisation.')


def _import_functions(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "code": {
                "fr": cell[1].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[4].decode('latin-1'),
                },
            "description": {
                "fr": cell[7].decode('latin-1'),
                "en": cell[8].decode('latin-1'),
                "nl": cell[2].decode('latin-1'),
                }
            }
        terms.append(term)
    key = KEY_STORAGE + '.function'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported functions.')


def _import_contacts(cells, relex_web):
    ct_terms = []
    pb_terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "lastname": cell[4].decode('latin-1'),
            "firstname": cell[8].decode('latin-1'),
            "phone": cell[6].decode('latin-1'),
            "email": cell[7].decode('latin-1'),
            "organisation": cell[10].decode('latin-1'),
            "cell": cell[11].decode('latin-1'),
            "function": cell[9].decode('latin-1'),
            }
        if cell[12] == 'CT':
            ct_terms.append(term)
        else:
            del term["function"]
            pb_terms.append(term)
    key = KEY_STORAGE + '.contact'
    setattr(relex_web, key, json.dumps(ct_terms))
    print('Imported contacts.')
    key = KEY_STORAGE + '.brusselspartners'
    setattr(relex_web, key, json.dumps(pb_terms))
    print('Imported brussels partners.')


# ================= Keywords Themes  =====================

def _import_keywords(cells, relex_web):
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "name": {
                "fr": cell[2].decode('latin-1'),
                "en": cell[6].decode('latin-1'),
                "nl": cell[3].decode('latin-1'),
                },
            }
        terms.append(term)
    key = KEY_STORAGE + '.keywords'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported keywords.')


def _import_themes(cells, cells_keywords, relex_web):
    cells_keywords_dict = {}
    for row in cells_keywords:
        if row[1] not in cells_keywords_dict.keys():
            cells_keywords_dict[row[1]] = [row[0]]
        else:
            cells_keywords_dict[row[1]].append(row[0])
    terms = []
    for cell in cells:
        term = {
            "id": cell[0].decode('latin-1'),
            "keywords": cells_keywords_dict.get(cell[0], []),
            "name": {
                "fr": cell[1].decode('latin-1'),
                "en": cell[3].decode('latin-1'),
                "nl": cell[2].decode('latin-1'),
                },
            }
        terms.append(term)
    key = KEY_STORAGE + '.theme'
    setattr(relex_web, key, json.dumps(terms))
    print('Imported themes.')


# ================= Projects  =====================

def _import_projects(csv_dict, relex_web):
    """
       PROJECTS:
           0: ID
           1: PARENT
           2: CODE
           3: DESC_NL
           4: DESC_EN
           5: INTERNETWORK => ORGANISATION TYPE
           6: URL
           7: CONTENT_NL
           8:
           9:
           10: START
           11: COMMENTS
           12:
           13: CONTENT_FR
           14:
           15: END
           16: DESC_FR
           17: DELETE
           18: PRJS_ID => STATUS (17: ACTIF, 18: INACTIF, 19: ARCHIVE)
           19: RELT_ID => RELATIONS (17: BILATERAL, 18: MULTILATERAL)
           20: CONTENT_EN
    """
    countries = {}
    for cell in csv_dict['PROJECTS_COUNTRIES']:
        if cell[0] not in countries.keys():
            countries[cell[0]] = []
        countries[cell[0]].append(cell[5])
    regions = {}
    for cell in csv_dict['PROJECTS_REGIONS']:
        if cell[0] not in regions.keys():
            regions[cell[0]] = []
        regions[cell[0]].append(cell[1])
    cities = {}
    for cell in csv_dict['PROJECTS_CITIES']:
        if cell[0] not in cities.keys():
            cities[cell[0]] = []
        cities[cell[0]].append(cell[5])
    contacts_type = {}
    for cell in csv_dict['CONTACTS']:
        contacts_type[cell[0]] = cell[12]
    contacts = {}
    brusselspartners = {}
    for cell in csv_dict['PROJECTS_CONTACTS']:
        if contacts_type[cell[5]] == 'PB':
            contacts_dict = brusselspartners
        else:
            contacts_dict = contacts
        if cell[0] not in contacts_dict.keys():
            contacts_dict[cell[0]] = []
        contacts_dict[cell[0]].append(cell[5])
    projects = {}
    for cell in csv_dict['PROJECTS']:
        if cell[1] not in projects.keys():
            projects[cell[1]] = []
        projects[cell[1]].append(cell)

    _create_projects(countries, regions, cities, contacts, brusselspartners,
                     projects, relex_web, '')
    print('Imported projects.')


def _create_projects(countries, regions, cities, contacts, brusselspartners,
                     projects, context, parent):
    for cell in projects.get(parent, []):
        # Deleted projects
        if cell[17].decode('latin-1') == 'O':
            continue
        project = _create_project(cell, context)
        project.setCountries(countries.get(cell[0], []))
        project.setRegions(regions.get(cell[0], []))
        project.setCities(cities.get(cell[0], []))
        project.setContacts(contacts.get(cell[0], []))
        project.setBrusselspartners(brusselspartners.get(cell[0], []))
        project.setTitleFromData()
        project.reindexObject()
        # Create children projects
        _create_projects(
            countries, regions, cities, contacts, brusselspartners,
            projects, project, cell[0]
        )


def _create_project(cell, context):
    STATUS = {'17': 'active', '18': 'inactive', '19': 'archive'}
    chooser = INameChooser(context)
    project_id = chooser.chooseName(cell[2].decode('latin-1'), context)

    # Create project
    context.invokeFactory("Project", project_id)
    project = context[project_id]
    project.setExcludeFromNav(True)

    # Set project attributes
    project.setCode(cell[2].decode('latin-1'))
    project.setName_fr(cell[16].decode('latin-1'))
    project.setName_en(cell[4].decode('latin-1'))
    project.setName_nl(cell[3].decode('latin-1'))
    project.setContent_fr(cell[13].decode('latin-1'))
    project.setContent_en(cell[20].decode('latin-1'))
    project.setContent_nl(cell[7].decode('latin-1'))
    project.setComments(cell[11].decode('latin-1'))
    project.setUrl(cell[6].decode('latin-1'))
    project.setOrganisationtype(cell[5].decode('latin-1'))
    project.setStatus(STATUS.get(cell[18].decode('latin-1'), ''))
    project.setRelationtype(cell[19].decode('latin-1'))
    project.setStart(_get_datetime(cell[10].decode('latin-1')))
    project.setEnd(_get_datetime(cell[15].decode('latin-1')))

    return project


def _get_datetime(date):
    if not date:
        return None
    day = date[0:2]
    month = date[3:5]
    year = date[6:8]
    if int(year) < 80:
        year = u'20' + year
    else:
        year = u'19' + year
    return DateTime('{}{}{}'.format(year, month, day))


# ================= Main method  =====================

def import_csv(relex_web, folder):
    csv_dict = get_csv_dict(folder)
    _import_continents(csv_dict['CONTINENTS'], relex_web)
    _import_countries(csv_dict['COUNTRIES'], relex_web)
    _import_regions(csv_dict['REGIONS'], relex_web)
    _import_cities(csv_dict['CITIES'], relex_web)

    _import_cells(csv_dict['CELLS'], relex_web)
    _import_relation_types(csv_dict['RELATIONTYPES'], relex_web)
    _import_organisation_types(csv_dict['ORGANIZATIONTYPES'], relex_web)
    _import_organisations(
        csv_dict['ORGANIZATIONS'], csv_dict['CELLS_ORAGANIZATIONS'], relex_web
        )
    _import_functions(csv_dict['FUNCTIONS'], relex_web)
    _import_contacts(csv_dict['CONTACTS'], relex_web)

    _import_keywords(csv_dict['KEYWORDS'], relex_web)
    _import_themes(
        csv_dict['THEMES'], csv_dict['KEYWORDS_THEMES'], relex_web
        )
    _import_projects(csv_dict, relex_web)


if 'app' in locals():
    relex_web = app.restrictedTraverse('Plone').restrictedTraverse('relex_web')
    folder = 'Relex'
    setSite(app.restrictedTraverse('Plone'))
    import_csv(relex_web, folder)

    relex_web._p_changed = 1
    # Commit transaction
    import transaction
    transaction.commit()
    # Perform ZEO client synchronization (if running in clustered mode)
    app._p_jar.sync()
